File: leetcode/304-range-sum-query-2d-immutable/main.go
package main

import "fmt"

/*
	1st attempt
	- similar approach to the leetcode 303
	Time of Constructor O(n)
	Space of Constructor O(n)
	Time of SumRange O(n)
	Space of SumRange O(1)
	52ms beats 80%
*/
type NumMatrix struct {
	Arr [][]int
}

func Constructor(matrix [][]int) NumMatrix {
	arr := [][]int{}
	for i := 0; i < len(matrix); i++ {
		temp := []int{}
		sum := 0
		for j := 0; j < len(matrix[i]); j++ {
			sum += matrix[i][j]
			temp = append(temp, sum)
		}
		arr = append(arr, temp)
	}
	return NumMatrix{arr}
}

func (this *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
	result := 0
	for i := row1; i <= row2; i++ {
		if col1 == 0 {
			result += this.Arr[i][col2]
		} else if col1 <= col2 {
			result += this.Arr[i][col2] - this.Arr[i][col1-1]
		}
	}
	return result
}

func main() {
	m := [][]int{
		[]int{3, 0, 1, 4, 2},
		[]int{5, 6, 3, 2, 1},
		[]int{1, 2, 0, 1, 5},
		[]int{4, 1, 0, 1, 7},
		[]int{1, 0, 3, 0, 5},
	}
	c := Constructor(m)
	fmt.Println(c.SumRegion(2, 1, 4, 3))
	fmt.Println(c.SumRegion(1, 1, 2, 2))
	fmt.Println(c.SumRegion(1, 2, 2, 4))
}

File: leetcode/304-range-sum-query-2d-immutable/main.py
"""
    1st: prefix sum on each row

    Time of init()      O(RC)
    Time of sumRage()   O(R)
    Space               O(RC)
    212 ms, faster than 6.65%
"""
class NumMatrix(object):

    def __init__(self, matrix):
        """
        :type matrix: List[List[int]]
        """
        self.cache = []
        for nums in matrix:
            pfs = 0
            pfss = []
            for x in nums:
                pfs += x
                pfss.append(pfs)
            self.cache.append(pfss)
        

    def sumRegion(self, row1, col1, row2, col2):
        """
        :type row1: int
        :type col1: int
        :type row2: int
        :type col2: int
        :rtype: int
        """
        total = 0
        for i in range(row1, row2 + 1):
            s = self.cache[i][col2]
            if col1-1 >= 0:
                s -= self.cache[i][col1-1]
            total += s
        return total


"""
    2nd: prefix sum + calculation

     a |  b  |
    ___|_____|__
       |     |
     c |  d  |
    ___|_____|__
       |     |
    
    Sum(ABCD) = Sum(OD) − Sum(OB) − Sum(OC) + Sum(OA)

    see ./idea.png

    Time of init()      O(RC)
    Time of sumRage()   O(1)
    Space               O(RC)
    96 ms, faster than 55.66%
"""

class NumMatrix(object):

    def __init__(self, matrix):
        """
        :type matrix: List[List[int]]
        """
        if not matrix or not matrix[0]:
            return
        R, C = len(matrix), len(matrix[0])
        cache = [C * [0] for _ in range(R)]
        for i in range(R):
            for j in range(C):
                top = cache[i-1][j] if i-1 >= 0 else 0
                left = cache[i][j-1] if j-1 >= 0 else 0
                topLeft = cache[i-1][j-1] if i-1 >= 0 and j-1 >= 0 else 0
                cache[i][j] = matrix[i][j] + top + left - topLeft
        self.cache = cache
        

    def sumRegion(self, row1, col1, row2, col2):
        """
        :type row1: int
        :type col1: int
        :type row2: int
        :type col2: int
        :rtype: int
        """
        if not self.cache:
            return 0
        top = self.cache[row1-1][col2] if row1-1 >= 0 else 0
        left = self.cache[row2][col1-1] if col1-1 >= 0 else 0
        topLeft = self.cache[row1-1][col1-1] if row1-1 >= 0 and col1-1 >= 0 else 0
        return self.cache[row2][col2] - top - left + topLeft